posthog/clickhouse/migrations/0137_session_replay_events_retention_period_days.py

```diff
@@ -0,0 +1,38 @@
+from posthog.clickhouse.client.connection import NodeRole
+from posthog.clickhouse.client.migration_tools import run_sql_with_exceptions
+from posthog.session_recordings.sql.session_replay_event_migrations_sql import (
+    ADD_RETENTION_PERIOD_DAYS_DISTRIBUTED_SESSION_REPLAY_EVENTS_TABLE_SQL,
+    ADD_RETENTION_PERIOD_DAYS_SESSION_REPLAY_EVENTS_TABLE_SQL,
+    ADD_RETENTION_PERIOD_DAYS_WRITABLE_SESSION_REPLAY_EVENTS_TABLE_SQL,
+    DROP_KAFKA_SESSION_REPLAY_EVENTS_TABLE_SQL,
+    DROP_RETENTION_PERIOD_DISTRIBUTED_SESSION_REPLAY_EVENTS_TABLE_SQL,
+    DROP_RETENTION_PERIOD_SESSION_REPLAY_EVENTS_TABLE_SQL,
+    DROP_RETENTION_PERIOD_WRITABLE_SESSION_REPLAY_EVENTS_TABLE_SQL,
+    DROP_SESSION_REPLAY_EVENTS_TABLE_MV_SQL,
+)
+from posthog.session_recordings.sql.session_replay_event_sql import (
+    KAFKA_SESSION_REPLAY_EVENTS_TABLE_SQL,
+    SESSION_REPLAY_EVENTS_TABLE_MV_SQL,
+)
+
+operations = [
+    # looking at past migrations we have to drop materialized views and kafka tables first
+    run_sql_with_exceptions(DROP_SESSION_REPLAY_EVENTS_TABLE_MV_SQL(on_cluster=False)),
+    run_sql_with_exceptions(DROP_KAFKA_SESSION_REPLAY_EVENTS_TABLE_SQL(on_cluster=False)),
+    # alter the target tables
+    run_sql_with_exceptions(ADD_RETENTION_PERIOD_DAYS_WRITABLE_SESSION_REPLAY_EVENTS_TABLE_SQL()),
+    run_sql_with_exceptions(
+        ADD_RETENTION_PERIOD_DAYS_DISTRIBUTED_SESSION_REPLAY_EVENTS_TABLE_SQL(),
+        node_roles=[NodeRole.DATA, NodeRole.COORDINATOR],
+    ),
+    run_sql_with_exceptions(ADD_RETENTION_PERIOD_DAYS_SESSION_REPLAY_EVENTS_TABLE_SQL(), sharded=True),
+    run_sql_with_exceptions(DROP_RETENTION_PERIOD_WRITABLE_SESSION_REPLAY_EVENTS_TABLE_SQL()),
```

Comment:
do we risk there being ingestion pods trying to write to these columns as we're dropping them?
might be safer to get this change out without the drop, so that all ingestion pods are using the new column and then drop in a follow-up

??

Reply:
After ingestion from Kafka to ClickHouse resumes when the Kafka table is recreated, there will probably be some messages in Kafka with the old field present and the new field not yet present.

I assume ClickHouse will ignore that 'extra' field and will set the new column to null which is fine.

Shortly after that the messages in Kafka will have just the new field present and we're off to the races.

Reply:
🤞